pkg/aws/eni/node.go

```diff
@@ -494,9 +495,9 @@ func (n *Node) ResyncInterfacesAndIPs(ctx context.Context, scopedLog *logrus.Ent
 
 // GetMaximumAllocatableIPv4 returns the maximum amount of IPv4 addresses
 // that can be allocated to the instance
-func (n *Node) GetMaximumAllocatableIPv4() int {
+func (n *Node) GetMaximumAllocatableIPv4(scopedLog *logrus.Entry) int {
```

Comment:
Why was this needed? It seems to me that we can already access the logger via `n.loggerLocked()`

Reply:
Plus this modifies the interface to the functions for no concrete benefit

Reply:
So this is exactly where the whole situation resides. By calling `n.loggerLocked()` within this scope, the function gets blocked as it returns `n.node.InstanceID()` which requires a read-lock over the node mutex which is already rw locked

https://github.com/cilium/cilium/blob/e2676483a51b7b089cd6a476b83038b9a1ec6169/pkg/aws/eni/node.go#L76

https://github.com/cilium/cilium/blob/e2676483a51b7b089cd6a476b83038b9a1ec6169/pkg/ipam/node.go#L311

https://github.com/cilium/cilium/blob/e2676483a51b7b089cd6a476b83038b9a1ec6169/pkg/ipam/node.go#L364

do you know if there is a better way to workaround this issue?

Reply:
Are you saying that you found a deadlock in the code prior and this change fixes it? Because the prior code is still using `loggerLocked()`

Reply:
yes, absolutely!

Reply:
Ah, I see! I didn't notice that the driver behind this PR is this deadlock. Nice catch!

However, I still think we can do better here than to pollute the interface with a logger; that shouldn't need to be passed around and the fact that do rely on that to avoid a deadlock is a very strong indicator that we need to rethink the flow and dependencies of this code.

Here's my proposal:

Let's add `instanceID string` as a field to the `pkg/aws/eni.Node` struct. This would allow us to avoid depending on `n.node.InstanceID()` which was done previously (which evidently cannot be called when `n.node` is already locked).

We can initialize the field when we call [`CreateNode`](https://github.com/mvisonneau/cilium/blob/e2676483a51b7b089cd6a476b83038b9a1ec6169/pkg/aws/eni/instances.go#L68), and we can rely on this because the instance ID is [always set when the resource is created](https://github.com/mvisonneau/cilium/blob/e2676483a51b7b089cd6a476b83038b9a1ec6169/pkg/nodediscovery/nodediscovery.go#L469).

Something like:

```
func (m *InstancesManager) CreateNode(obj *v2.CiliumNode, n *ipam.Node) ipam.NodeOperations {
	return &Node{k8sObj: obj, manager: m, node: n, instanceID: n.InstanceID()}
}
```

or feel free to create a constructor for `Node`. 

If you think this is a good idea, then please implement in a separate commit. WDYT?

Reply:
indeed, this is much cleaner and future proof, thanks for the guidance on this @christarazi 🙇 !